x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -449,33 +449,36 @@ Entry<CacheFileRegion> get(KeyType cacheKey, long fileLength, int region) {
      * @param cacheKey  the key to fetch data for
      * @param length    the length of the blob to fetch
      * @param writer    a writer that handles writing of newly downloaded data to the shared cache
+     * @param listener  listener that is called once all downloading has finished
      *
      * @return {@code true} if there were enough free pages to start downloading
      */
-    public boolean maybeFetchFullEntry(KeyType cacheKey, long length, RangeMissingHandler writer) {
+    public boolean maybeFetchFullEntry(KeyType cacheKey, long length, RangeMissingHandler writer, ActionListener<Void> listener) {
         int finalRegion = getEndingRegion(length);
         if (freeRegionCount() < finalRegion) {
             // Not enough room to download a full file without evicting existing data, so abort
             return false;
         }
         long regionLength = regionSize;
-        for (int region = 0; region <= finalRegion; region++) {
-            var entry = get(cacheKey, length, region);
-            if (region == finalRegion) {
-                regionLength = length - getRegionStart(region);
-            }
-            ByteRange rangeToWrite = ByteRange.of(0, regionLength);
-            if (rangeToWrite.length() == 0) {
-                return true;
+        try (RefCountingListener refCountingListener = new RefCountingListener(listener)) {
+            for (int region = 0; region <= finalRegion; region++) {
+                var entry = get(cacheKey, length, region);
+                if (region == finalRegion) {
+                    regionLength = length - getRegionStart(region);
+                }
+                ByteRange rangeToWrite = ByteRange.of(0, regionLength);
+                if (rangeToWrite.isEmpty()) {
+                    return true;
+                }
+                entry.chunk.populateAndRead(
+                    rangeToWrite,
+                    ByteRange.EMPTY,
+                    (channel, pos, relativePos, len) -> 0,
+                    writer,
+                    bulkIOExecutor,
+                    refCountingListener.acquire(ignored -> {})
```

Comment:
I think this listener is called after the read, not after the write?  So we actually need another listener that is called when the gaps complete?

Reply:
🤦 yea sorry about that, I fixed this now in https://github.com/elastic/elasticsearch/pull/99194/commits/40b249ca169b2620d4ee02cc9c78fb1d5ab00dd5 by just making read and write range the same, that should do the trick here without requiring any bigger changes. Since we're not actually doing any reading that should be fine I think.

Reply:
Yep, that should work AIUI.  Might be worth adding a comment to that effect?

Reply:
++ added thanks Alan!